### PR TITLE
Improve README

### PR DESCRIPTION
**Pull request by miloradvojnovic**
Summary:
 - Improved README with more examples, screenshots
 - Linked Contributing Guide and added instructions for a local publishing

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,7 +7,11 @@
 ![logo_ash_horizontal@4x](https://user-images.githubusercontent.com/25779550/135059025-4cfade5b-bfcb-47e8-872f-8a3d78ce0c25.png)
 
 Serialization toolbox for Akka messages, events and persistent state that helps achieve compile-time guarantee on
-serializability.
+serializability. There're some Akka serialization caveats this tool can help with:
+1. [Missing serialization binding](#missing-serialization-binding)
```
Comment (PawelLipski) [inline README.md:11]: Yeah, makes more sense to organize the README around **caveats**, just as in the sphere.it preso

File: README.md [hunk 2]
```diff
@@ -7,7 +7,11 @@
...
+2. [Inconsistent persistent data](#inconsistent-persistent-data)
```
Comment (PawelLipski) [inline README.md:12]: Maybe:
```suggestion
2. [Incompatibility of persistent data](#...)
```
❓

File: README.md [hunk 3]
```diff
@@ -23,16 +27,42 @@ lazy val app = (project in file("app"))
   .enablePlugins(AkkaSerializationHelperPlugin)
 ```
 
-## Features
+## Missing serialization binding
 
-### 1. Serializability Checker
+To serialize message, state or event in Akka, Scala trait needs to be defined:
```
Comment (PawelLipski) [inline README.md:32]:
```suggestion
To serialize message, persistent state or event in Akka, Scala trait needs to be defined:
```

As _state_ alone is a bit unclear/imprecise (I wouldn't think of akka-persistence first)

Comment (miloradvojnovic) [reply README.md:32]: Definitely. It's more meaningful. I would apply this suggestion.

File: README.md [hunk 4]
```diff
@@ -23,16 +27,42 @@ lazy val app = (project in file("app"))
...
+  serializers {
+    jackson-json = "akka.serialization.jackson.JacksonJsonSerializer"
+  }
+  serialization-bindings {
+    "org.MySer" = jackson-json
+  }
+}
+```
+
+The problem occurs if we forget to use created serializer:
```
Comment (PawelLipski) [inline README.md:52]:
```suggestion
The problem occurs if we forget to extend a class with the base trait bound to the serializer:
```

File: README.md [hunk 5]
```diff
@@ -51,19 +81,23 @@ object BehaviorTest {
...
+test0.scala:6: error: Make sure this type is itself annotated, or extends a type annotated
+with  @org.virtuslab.ash.annotation.SerializabilityTrait.
   sealed trait Command extends MySerializable
                ^
 ```
 
+
 The compiler plugin only checks the classes in the sbt modules where `AkkaSerializationHelperPlugin` is explicitly enabled.
-It may happen that your base trait (like `MySerializable` in the example) lives in an sbt module like `core` where you do **not** want to enable the plugin
+It may happen that your base trait (like `MySerializable` in the example) may live in an sbt module like `core` where you do **not** want to enable the plugin
```
Comment (PawelLipski) [inline README.md:100]: No need for double `may` (`may happen ... may live`), pls revert the change

File: README.md [hunk 6]
```diff
@@ -76,9 +110,14 @@ lazy val core = (project in file("core"))
   .settings(libraryDependencies += AkkaSerializationHelperPlugin.annotation)
 ```
 
-### 2. Dump Persistence Schema
+## Incompatibility of persistent data
+
+![Typical tragic story](docs/typical-tragic-story.png "Typical tragic story")
```
Comment (PawelLipski) [inline README.md:115]: Nitpick: doesn't the text in `[...]` already serve as alt-text? is the text in `"..."` also needed?

File: README.md [hunk 7]
```diff
@@ -76,9 +110,14 @@ lazy val core = (project in file("core"))
...
 
-A mix of a compiler plugin and an sbt task for dumping schema
+A typical problem we could have with persistence is when we persisted some data that is not compatible
```
Comment (PawelLipski) [inline README.md:117]:
```suggestion
A typical problem we could have with persistence is when the already persisted data is not compatible
```

File: README.md [hunk 8]
```diff
@@ -76,9 +110,14 @@ lazy val core = (project in file("core"))
...
+with a new version.
```
Comment (PawelLipski) [inline README.md:118]:
```suggestion
with the schemas defined in a new version of the application.
```

File: README.md [hunk 9]
```diff
@@ -51,19 +81,23 @@ object BehaviorTest {
...
   sealed trait Command extends MySerializable
                ^
 ```
 
+
 The compiler plugin only checks the classes in the sbt modules where `AkkaSerializationHelperPlugin` is explicitly enabled.
-It may happen that your base trait (like `MySerializable` in the example) lives in an sbt module like `core` where you do **not** want to enable the plugin
+It may happen that your base trait (like `MySerializable` in the example) may live in an sbt module like `core` where you do **not** want to enable the plugin
 (e.g. for compilation performance reasons).
 You still need to reference `org.virtuslab.ash.SerializabilityTrait` to annotate `MySerializable`, however.
```
Comment (PawelLipski) [inline README.md:102]: The usage of `you` and `we` is inconsistent across the text.

I'd prefer using impersonal form wherever natural, or `you` as the second choice; refrain from using `we` as it sounds somewhat... idk, immature (?) in technical writing

Comment (PawelLipski) [reply README.md:102]: (esp. that it's unclear to whom `we` refers: _we the authors_? _we = the author + the reader_?)

File: README.md [hunk 10]
```diff
@@ -118,10 +157,64 @@ ashDumpPersistenceSchemaOutputDirectoryPath := "~" // Changes directory
...
+There will not be exceptions during serialization but Jackson will create
+another instance of Tick instead of restoring Singleton.
+
+```scala
+actorRef ! Tick
+
+// Inside the actor:
+def receive = {
+  case Tick => // this won't get matched
+}// message will be unhandled
```
Comment (PawelLipski) [inline README.md:213]: Nitpick:
```suggestion
} // message will be unhandled
```

File: README.md [hunk 11]
```diff
@@ -164,13 +257,15 @@ read [Akka documentation about serialization](https://doc.akka.io/docs/akka/curr
 [`CirceAkkaSerializer`](https://github.com/VirtusLab/akka-serialization-helper/blob/main/circe-akka-serializer/src/main/scala/org/virtuslab/ash/circe/CirceAkkaSerializer.scala) Scaladoc
 and look at the [examples](https://github.com/VirtusLab/akka-serialization-helper/tree/main/examples).
 
-### 4. Codec Registration Checker
+## Missing Codec registration
 
-Compiler plugin for checking whether all codecs are registered.
-During compilation, it gathers all direct descendants of the class marked with `@org.virtuslab.ash.SerializabilityTrait`
-and checks the body of classes annotated with `@org.virtuslab.ash.Serializer` if they reference all these direct descendants in any way.
+If we forget to register a codec we would have problems in run time. To solve that we can use
```
Comment (PawelLipski) [inline README.md:262]: `runtime` is typically written as a single world (at least from what I've seen in Scala contexts)

File: README.md [hunk 12]
```diff
@@ -164,13 +257,15 @@ read [Akka documentation about serialization](https://doc.akka.io/docs/akka/curr
...
+annotation [`@org.virtuslab.ash.Serializer`](https://github.com/VirtusLab/akka-serialization-helper/blob/main/annotation/src/main/scala/org/virtuslab/ash/annotation/Serializer.scala)
```
Comment (PawelLipski) [inline README.md:263]:
```suggestion
annotation [`@org.virtuslab.ash.Serializer`](https://github.com/VirtusLab/akka-serialization-helper/blob/main/annotation/src/main/scala/org/virtuslab/ash/annotation/Serializer.scala).
```

File: README.md [hunk 13]
```diff
@@ -217,15 +312,18 @@ Test / ashCompilerPluginEnable := false
 
 For full list of sbt keys, check [`org.virtuslab.ash.AkkaSerializationHelperKeys`](https://github.com/VirtusLab/akka-serialization-helper/blob/main/sbt-akka-serialization-helper/src/main/scala/org/virtuslab/ash/AkkaSerializationHelperKeys.scala).
 
+## Contributing Guide
+
+If you want to contribute to the ASH checkout our [Contributing Guide](https://github.com/VirtusLab/akka-serialization-helper/blob/main/CONTRIBUTING.md)
```
Comment (PawelLipski) [inline README.md:317]:
```suggestion
If you want to contribute to this project, see [Contributing Guide](https://github.com/VirtusLab/akka-serialization-helper/blob/main/CONTRIBUTING.md)
```

File: README.md [hunk 14]
```diff
@@ -36,7 +36,7 @@ package org
 trait MySer
 ```
 
-Also, we need to define a serializer for the Akka in the Scala configuration file:
+Also, a serializer needs to be defined for the Akka in the Scala configuration file:
```
Comment (PawelLipski) [inline README.md:39]:
```suggestion
Also, a serializer needs to be defined for Akka in Scala configuration file:
```

Comment (PawelLipski) [reply README.md:39]: Generally, no need to precede proper names with _the_ article in English

File: README.md [hunk 15]
```diff
@@ -49,27 +49,27 @@ akka.actor {
...
-The problem occurs if we forget to use created serializer:
+The problem occurs if a class is not extended with the base trait bound to the serializer:
 
 ```scala
 trait MySer
 case class MyMessage() // extends MySer
 ```
 
-To solve that we can use this plugin that detects messages, events and persistent states, and checks whether they
+To solve that this plugin can be used. It detects messages, events and persistent states, and checks whether they
```
Comment (PawelLipski) [inline README.md:59]:
```suggestion
akka-serialization-helper to the rescue! It detects messages, events and persistent states, and checks whether they
```

File: README.md [hunk 16]
```diff
@@ -97,11 +97,11 @@ with  @org.virtuslab.ash.annotation.SerializabilityTrait.
 
 
 The compiler plugin only checks the classes in the sbt modules where `AkkaSerializationHelperPlugin` is explicitly enabled.
-It may happen that your base trait (like `MySerializable` in the example) may live in an sbt module like `core` where you do **not** want to enable the plugin
-(e.g. for compilation performance reasons).
-You still need to reference `org.virtuslab.ash.SerializabilityTrait` to annotate `MySerializable`, however.
+It may happen that the base trait (like `MySerializable` in the example) lives in a sbt module like `core` where the plugin
```
Comment (PawelLipski) [inline README.md:100]:
```suggestion
It may happen that the base trait (like `MySerializable` in the example) lives in an sbt module like `core` where the plugin
```
since it's `es-bee-tee` in pronunciation

File: README.md [hunk 17]
```diff
@@ -97,11 +97,11 @@ with  @org.virtuslab.ash.annotation.SerializabilityTrait.
...
+should **not** be enabled (e.g. for compilation performance reasons).
+However, `MySerializable` needs to be annotated with a `org.virtuslab.ash.SerializabilityTrait`.
```
Comment (PawelLipski) [inline README.md:102]:
```suggestion
However, `MySerializable` needs to be annotated with `org.virtuslab.ash.SerializabilityTrait`.
```
Again, article should not precede a proper name

File: README.md [hunk 18]
```diff
@@ -97,11 +97,11 @@ with  @org.virtuslab.ash.annotation.SerializabilityTrait.
...
 In order to have access to the `SerializabilityTrait` annotation without enabling the entire suite of compiler plugins,
-add `AkkaSerializationHelperPlugin.annotation` to `libraryDependencies`:
```
Comment (PawelLipski) [inline README.md:70]: Imperative is okay! probably even better than passive voice, which sometimes sounds too formal/stiff

File: README.md [hunk 19]
```diff
@@ -112,12 +112,12 @@ lazy val core = (project in file("core"))
...
-![Typical tragic story](docs/typical-tragic-story.png "Typical tragic story")
+![Typical tragic story](docs/typical-tragic-story.png)
 
-A typical problem we could have with persistence is when we persisted some data that is not compatible
-with a new version.
+A typical problem with a persistence is when the already persisted data is not compatible
+with the schemas defined in a new version of the application.
 
-To solve this we can use a mix of a compiler plugin and an sbt task for dumping schema
+To solve this a mix of a compiler plugin and a sbt task for dumping schema can be used for dumping schema
```
Comment (PawelLipski) [inline README.md:120]:
```suggestion
To solve this, a mix of a compiler plugin and an sbt task can be used for dumping schema
```

File: README.md [hunk 20]
```diff
@@ -157,10 +157,10 @@ ashDumpPersistenceSchemaOutputDirectoryPath := "~" // Changes directory
   - org.random.project.Data
 ```
 
-We can use diff to check the difference between the last version of a schema and the
+A `diff` command can be used to check the difference between the last version of a schema and the
```
Comment (PawelLipski) [inline README.md:160]:
```suggestion
A `diff` command can be used to check the difference between the version of a schema from `develop`/`main` branch and the
```

File: README.md [hunk 21]
```diff
@@ -179,7 +179,7 @@ final case class Lion(name: String) extends Animal
 final case class Tiger(name: String) extends Animal
 ```
 
-We need to add a lot of Jackson annotations to run this code:
+To run this code a lot of Jackson annotations should be added:
```
Comment (PawelLipski) [inline README.md:182]:
```suggestion
To run this code, a lot of Jackson annotations should be added:
```

File: README.md [hunk 22]
```diff
@@ -259,8 +259,9 @@ and look at the [examples](https://github.com/VirtusLab/akka-serialization-helpe
 
 ## Missing Codec registration
 
-If we forget to register a codec we would have problems in run time. To solve that we can use
-annotation [`@org.virtuslab.ash.Serializer`](https://github.com/VirtusLab/akka-serialization-helper/blob/main/annotation/src/main/scala/org/virtuslab/ash/annotation/Serializer.scala)
+If a codec is not registered there will be problems in a runtime. To solve that an annotation
```
Comment (PawelLipski) [inline README.md:262]:
```suggestion
If a codec is not registered, there will be problems in a runtime. To solve that an annotation
```

Comment (PawelLipski) [reply README.md:262]: Pls specify what problems, on the example of Circe serializer

**Review by PawelLipski — COMMENTED** [review]
A bunch of language comments... also, in your spare time, see https://virtuslab.github.io/tips/#english

File: README.md [hunk 23]
```diff
@@ -259,7 +259,9 @@ and look at the [examples](https://github.com/VirtusLab/akka-serialization-helpe
 
 ## Missing Codec registration
 
-If a codec is not registered there will be problems in a runtime. To solve that an annotation
+If a codec is not registered, there will be problems in a runtime. This will lead to a missing
+codec for Akka serialization. It's one of the main caveats of using
+a serializer that depends on the compile-time. To solve that, an annotation
```
Comment (PawelLipski) [inline README.md:264]: `depends on the compile-time`? 🤔

File: CONTRIBUTING.md [hunk 24]
```diff
@@ -19,6 +19,14 @@ To build docs, type
 sbt doc
 ```
 
+### Running
+If some changes are made we can test it by publishing it to the local Maven repository
```
Comment (PawelLipski) [inline CONTRIBUTING.md:23]:
```suggestion
If some changes are made, you can test them by publishing it to the local Maven repository
```

File: README.md [hunk 25]
```diff
@@ -23,23 +27,49 @@ lazy val app = (project in file("app"))
...
-trait and report an error when they don't. This ensures that the specified serializer is used by Akka and protects
-against an accidental use
-of [Java serialization](https://doc.akka.io/docs/akka/current/serialization.html#java-serialization) or outright serialization failure.
+```scala
+package org
+trait MySer
+```
 
-To use, annotate your base trait with `@org.virtuslab.ash.SerializabilityTrait`:
+Also, a serializer needs to be defined for Akka in Scala configuration file:
```
Comment (PawelLipski) [inline README.md:39]:
```suggestion
Also, a serializer needs to be bound to this trait in a configuration file:
```

Comment (PawelLipski) [reply README.md:39]: (HOCON-formatted config files aren't specific to Scala after all, their API is Java-compatible)

File: README.md [hunk 26]
```diff
@@ -23,23 +27,49 @@ lazy val app = (project in file("app"))
...
+The problem occurs if a class is not extended with the base trait bound to the serializer:
+
+```scala
+trait MySer
+case class MyMessage() // extends MySer
+```
+
+`akka-serialization-helper` to the rescue! It detects messages, events and persistent states, and checks whether they
+extend the given base trait and report an error when they don't. This ensures that the specified serializer is
+used by Akka and protects against accidental use of
```
Comment (PawelLipski) [inline README.md:61]:
```suggestion
used by Akka and protects against an unintended fallback to
```

File: README.md (hunk 18 above)
Comment (PawelLipski) [reply README.md:70]: So... pls revert here

File: README.md [hunk 27]
```diff
@@ -118,13 +157,67 @@ ashDumpPersistenceSchemaOutputDirectoryPath := "~" // Changes directory
...
-### 3. Serializer
+A `diff` command can be used to check the difference between the version of a schema from `develop`/`main` branch and the
+version from the current commit.
+
+![Easy to diff](docs/easy-to-diff.png)
+
+
+## Jackson Akka Serializer
+
+Using Jackson Serializer for Akka-persistence is also one of the pitfalls and this
```
Comment (PawelLipski) [inline README.md:168]:
```suggestion
Using Jackson Serializer for akka-persistence is also one of the pitfalls and this
```

File: README.md [hunk 28]
```diff
@@ -118,13 +157,67 @@ ashDumpPersistenceSchemaOutputDirectoryPath := "~" // Changes directory
...
+plugin provides an alternative by using a serializer that uses Circe.
```
Comment (PawelLipski) [inline README.md:169]: Pls link to Circe here

File: README.md [hunk 29]
```diff
@@ -118,13 +157,67 @@ ashDumpPersistenceSchemaOutputDirectoryPath := "~" // Changes directory
...
+```scala
+case class Message(animal: Animal) extends MySer
+
+sealed trait Animal
+
+final case class Lion(name: String) extends Animal
+final case class Tiger(name: String) extends Animal
+```
+
+To run this code, a lot of Jackson annotations should be added:
```
Comment (PawelLipski) [inline README.md:182]: `To run this code` is a bit imprecise, this code _will run_ without these annotations ;)

File: README.md [hunk 30]
```diff
@@ -118,13 +157,67 @@ ashDumpPersistenceSchemaOutputDirectoryPath := "~" // Changes directory
...
 
-[Circe-based](https://circe.github.io/circe/) Akka serializer. It uses Circe codecs, derived using [Shapeless](https://circe.github.io/circe/codecs/auto-derivation.html),
-that are generated during compile time (so serializer won't crash during runtime like reflection-based serializers may do).
+Also if an object is defined:
+```scala
+case object Tick
+```
+
+There will not be exceptions during serialization but Jackson will create
+another instance of Tick instead of restoring Singleton.
```
Comment (PawelLipski) [inline README.md:205]:
```suggestion
another instance of `Tick` instead of restoring the `object Tick`'s underlying singleton.
```

**Review by PawelLipski — APPROVED** [review]
